### ISSUE TITLE
Improve auto-casting in JSON::Any

### ISSUE THREAD

**Issue by al6x** [crystal-lang/crystal#8618]
Currently JSON is inconvenient to use, not sure if that's a bug or feature. The perfectly reasonable code below would fail complaining `Int` can't be cast to `Float`

``` ruby
require "json"
JSON.parse("{\"v\":1}")["v"].as_f
```

**Expected behaviour**: The code above should work and JSON::Any should allow reasonable automatic casting between `Int` -> `Float`, `Number` -> `String` etc.

How I hit this bug - I tried to parse JSON like `{ a: 1, b: 1.1 }` - when some values casted to `Int` and some to `Float` and the only way to handle it is to `json.to_i? || json.to_f` which feels wrong.

**Comment by Blacksmoke16** [comment]
FWIW a better approach would be to deserialize the data into structs/classes via `JSON::Serializable` or `JSON.mapping`.  This also works for literal types.

`Hash(String, Float64).from_json "{\"v\":1}"`

**Comment by alexeypetrushin** [comment]
In my opinion having `JSON::Any` is fine, it's easy to understand and use.

**Comment by RX14** [comment]
There should probably be `to_i` and `to_f` in addition to `as_i` and `as_f`, which are fuzzy on whether the value is an integer or float.

**Comment by asterite** [comment]
Why not use `as_i.to_f`? It makes it clear that you are converting an int to float, which might result in precision loss.

**Comment by RX14** [comment]
@asterite because that will fail if the value is `1.0`, since it's a json float.

@alexeypetrushin is asking how to treat `1` and `1.0` in json *both as floats*, for libraries which don't make the distinction.

**Comment by j8r** [comment]
There is no issue on behavior, the JSON document is represented correctly.

What is requested is to simplify casting, which is usually better to be explicit in my opinion.

Converting types to another can be already done by using `#raw`.
For example, converting all types to `String` will be `#raw.to_s`. In your case, you can use `#raw.as(Float64 | Int64).to_f`.

**Comment by asterite** [comment]
What is this API we are talking about? I don't understand why sometimes it's an int, sometimes a float. And the OP already provided a solution so I'm not sure there's something to do here.

**Comment by alexeypetrushin** [comment]
The `#raw.as(Float64 | Int64).to_f` should work.

@asterite I'm parsing historical price data, it's like `[{ date: "2020-01-01", price: 500 }, { date: "2020-01-02", price: 521.2 }, , ...] }` JSON parses numbers sometimes as `Int` and sometimes as `Float` which makes it hard to parse.

**Comment by ysbaddaden** [comment]
If the value is an explicit float (1.0)  then Any will parse it as a Float64, if it's ambiguous (1) the type will be an Int64, because it looks like an integer.

The `Any#as_i` is basically doing `.as(Int64)` and same for `as_f` (`.as(Float64)`). It's a basic type cast with no transformation. If Any parsed an Int64 (1) then trying to cast a Float64 out of the JSON::Type union will fail.

There are workarounds, as explained above, for example using manual casts with an explicit transformation of the raw union, like `.raw.as(Int64 | Float64).to_f`. That will always return a float, but that's not very pretty nor friendly to use :(

Maybe there could be explicit `Any#to_f` and `Any#to_i` methods that would do that for Float64 and Int64 (and maybe even String) automatically? I think it would fit into Any's purpose: parse whatever inefficiently but simply and nicely.

**Comment by asterite** [comment]
Sounds good. I guess `to_f` would call `to_f` on a string if the value is a string? Same for `to_i`? But then `to_i` returns `Int64` I guess?

Or maybe `as_f` could just cast the int to a float if it's an int, and that's it.

**Comment by straight-shoota** [comment]
> Or maybe `as_f` could just cast the int to a float if it's an int, and that's it.

No. `as_f` should *only* return the raw value (if it is a float).

Adding `Any#to_f` and `Any#to_i` sounds reasonable.

**Comment by lbarasti** [comment]
Just a note that whatever is decided above could (should?) be applied to `YAML::Any`, as well 🙏

**Comment by compumike** [comment]
This behavior hit me with a surprising exception in production when I was parsing data from an IoT device endpoint that sometimes looks like `{"voltage": 120.68, "power": 16.45}` but occasionally looks like `{"voltage": 121, "power": 16.53}` when the decimal part rounds away.

Same basic problem as @al6x's stock data use case described above. (I posted with more detail on https://forum.crystal-lang.org/t/json-any-as-f-unpredictable-when-number-might-sometimes-be-an-integer/5165 and @HertzDevil sent me here, thanks!)

My gut instinct is that this `#as_f` is something unexpected that will bite beginners. I think this parse-a-JSON-integer case is perhaps a special one, and should "just work" by default. Currently, it really depends on the source that wrote the JSON: some languages will emit `121.0` and others `121`.

If I may make a UX suggestion: one possibility I haven't seen discussed might be to add an argument that is enabled by default: `JSON::Any#as_f(cast_from_int : Bool = true)`.

This way, `JSON.parse("121").as_f` would work returning `121_f64`, while `JSON.parse("121").as_f(cast_from_int: false)` would maintain the current raw-only behavior and raise `TypeCastError`.

Whatever is decided, some additional documentation on `JSON::Any#as_f` about this issue would be nice.

**Comment by asterite** [comment]
Let's please make as_f also work with int values.

**Comment by straight-shoota** [comment]
Despite what I said earlier, I suppose that makes sense.

JSON only has a generic number type, so distinguishing between int and float is entirely internal to the Crystal mapping. The meaning of `#as_f` should not be influenced by the initial decision of the JSON parser to map the value to an int type. In JSON, it's just `Number`. If you want that to be represented as a float, you should get a float. JSON does not say it's *not a float*.

**Comment by asterite** [comment]
> JSON only has a generic number type

I think this is the key thing for deciding how to handle this.

**Comment by compumike** [comment]
PR #12835 is attached and updates `JSON::Any, YAML::Any` methods `#as_f, #as_f?, #as_f32, #as_f32?` to work when the underlying `@raw.is_a?(Int)`.
